keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,431 @@
+
+### Risks and Mitigations
+
+#### Cardinality growth of metrics
+
+The proposal will add O(pods*resource_types) series. Currently a larger number of series are exposed via `kube-state-metrics` to help summarize to this data, but this proposal should reduce the need for those metrics and lead to a cardinality reduction in the future. This proposal grows the number of per pod series by a small fraction of the existing amount, described in the [kube-state-metrics project docs](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md). In extremely large clusters an administrator may wish to not scrape or gather the metrics described here using the described flag.
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Expose new metrics
+
+The scheduler will expose an optional set of metrics series that capture the resources a pod is considered to be "consuming" from a scheduling perspective at the current time. The metrics will be exposed via `/metrics/resources` to allow aggregation to be optional and to allow a separate scrape interval. Only the active scheduler will report metrics series in order to remove the need to downsample the series and to reduce overall cardinality. If no scheduler is currently leading, then these metrics will temporarily not be reported which is consistent with metrics based on populated caches in leader elected components.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity. After a pod has reached the Completed lifecycle (as described above) state, the series will not be reported. If the value of the resource is zero, that resource will not report a series. They all share the form `kube_(running_|)pod_resource_(requests|limits)`.
```

Comment:
When a pod enters completed the series will be removed, so there will be no reported metric (we return null instead of 0 for completed pods).

Note that the implementation here doesn't use prometheus metrics in a registry, we use the higher performance "collector pattern" where a scrape triggers a traversal of the pod cache and each pod is converted to a metric on demand.  This requires no locking or other retention and is how we materialize most high cardinality metrics in the system from kubelet, kube-state-metrics, and things like node exporter.  So there is no such thing as a cached metric in this model.

Reply:
Will clarify that in the kep.